resolve_test.go

```diff
@@ -96,36 +104,53 @@ func TestSimpleIPResolve(t *testing.T) {
 	}
 }
 
-func TestResolveMultiple(t *testing.T) {
+func TestResolveOnlyOnce(t *testing.T) {
 	ctx := context.Background()
 	resolver := makeResolver()
 
 	addrs, err := resolver.Resolve(ctx, ma.StringCast("/dns4/example.com/quic/dns6/example.com"))
 	if err != nil {
 		t.Error(err)
 	}
+
 	for i, x := range []ma.Multiaddr{ip4ma, ip4mb} {
-		for j, y := range []ma.Multiaddr{ip6ma, ip6mb} {
-			expected := ma.Join(x, ma.StringCast("/quic"), y)
-			actual := addrs[i*2+j]
-			if !expected.Equal(actual) {
-				t.Fatalf("expected %s, got %s", expected, actual)
+		expected := ma.Join(x, ma.StringCast("/quic/dns6/example.com"))
+		actual := addrs[i]
+		if !expected.Equal(actual) {
+			t.Fatalf("expected %s, got %s", expected, actual)
+		}
+	}
+}
+
+func resolveAllDNS(ctx context.Context, resolver *Resolver, in ma.Multiaddr) ([]ma.Multiaddr, error) {
+	var inAddrs []ma.Multiaddr
+	outAddrs := []ma.Multiaddr{in}
+
+	for len(inAddrs) != len(outAddrs) {
+		inAddrs = outAddrs
+		outAddrs = nil
+		for _, inAddr := range inAddrs {
+			addrs, err := resolver.Resolve(ctx, inAddr)
+			if err != nil {
+				return nil, err
 			}
+			outAddrs = append(outAddrs, addrs...)
```

Comment:
This looks brittle. 
I guess you want to resolve *all* dns components here. Can you use the `Matches` function in utils? 

NOTE: I don't think we should be resolving all dns components, but only the ones at the front of the address.